processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -98,287 +167,248 @@ public void start()
     }
-    finally {
-      startStopLock.readLock().unlock();
-    }
-    return false;
   }
 
-  /**
-   * @param lookupName key to fetch the reference of the object {@link LookupExtractor}
-   *
-   * @return reference of {@link LookupExtractorFactory} that correspond the {@code lookupName} or null if absent
-   *
-   * @throws IllegalStateException if the {@link LookupReferencesManager} is closed or did not start yet
-   */
-  @Nullable
-  public LookupExtractorFactory get(String lookupName)
+  private void assertStarted()
   {
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      final LookupExtractorFactory lookupExtractorFactory = lookupMap.get(lookupName);
-      assertStarted();
-      return lookupExtractorFactory;
-    }
-    finally {
-      startStopLock.readLock().unlock();
+    if (!started) {
+      throw new ISE("LookupReferencesManager is not started.");
     }
   }
 
-  /**
-   * @return Returns {@link Map} containing a copy of the current state.
-   *
-   * @throws ISE if the is is closed or did not start yet.
-   */
-  public Map<String, LookupExtractorFactory> getAll()
+  @VisibleForTesting
+  interface Notice
   {
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      assertStarted();
-      return Maps.newHashMap(lookupMap);
-    }
-    finally {
-      startStopLock.readLock().unlock();
-    }
+    void handle();
   }
 
-  private void assertStarted() throws ISE
+  private class LoadNotice implements Notice
   {
-    if (isClosed()) {
-      throw new ISE("lookup manager is closed");
+    String lookupName;
+    LookupExtractorFactoryContainer lookupExtractorFactoryContainer;
+
+    public LoadNotice(String lookupName, LookupExtractorFactoryContainer lookupExtractorFactoryContainer)
+    {
+      this.lookupName = lookupName;
+      this.lookupExtractorFactoryContainer = lookupExtractorFactoryContainer;
     }
-  }
 
-  public boolean isClosed()
-  {
-    return !started.get();
+    @Override
+    public void handle()
+    {
+      startStopLock.readLock().lock();
+
+      try {
+        LookupExtractorFactoryContainer old = lookupMap.get(lookupName);
+        if (old != null && !lookupExtractorFactoryContainer.replaces(old)) {
+          LOG.warn(
+              "got notice to load lookup [%s] that can't replace existing [%s].",
+              lookupExtractorFactoryContainer,
+              old
+          );
+          return;
+        }
+      } finally {
+        startStopLock.readLock().unlock();
+      }
+
+      if (!lookupExtractorFactoryContainer.getLookupExtractorFactory().start()) {
+        throw new ISE("start method returned false for lookup [%s]:[%s]", lookupName, lookupExtractorFactoryContainer);
+      }
+
+      startStopLock.writeLock().lock();
```

Comment:
no, factory start is implemented in the extensions and can take arbitrarily long... that is why i have been extra careful in the LookupReferencesManager implementation to not hold any locks while you start/stop any factory.